tests/test_jax_nonbonded.py

```diff
@@ -0,0 +1,151 @@
+import jax
+jax.config.update("jax_enable_x64", True)
+
+import numpy as onp
+from numpy.random import randn, rand, randint, seed
+seed(2021)
+
+from jax import numpy as np, value_and_grad, jit, vmap
+
+from jax.ops import index_update, index
+from timemachine.potentials.nonbonded import nonbonded_v3, nonbonded_v3_on_specific_pairs
+from timemachine.potentials.jax_utils import convert_to_4d, get_all_pairs_indices, get_group_group_indices
+
+from functools import partial
+from typing import Tuple, Callable
+
+Conf = Params = Box = ChargeMask = LJMask = LambdaPlaneIdxs = LambdaOffsetIdxs = np.array
+Lamb = Beta = Cutoff = Energy = float
+
+nonbonded_args = Conf, Params, Box, Lamb, ChargeMask, LJMask, Beta, Cutoff, LambdaPlaneIdxs, LambdaOffsetIdxs
+NonbondedArgs = Tuple[nonbonded_args]
+NonbondedFxn = Callable[[*nonbonded_args], Energy]
+
+
+def generate_random_inputs(n_atoms: int, dim: int = 3) -> NonbondedArgs:
+    # jittered
+    offsets = onp.array([onp.arange(n_atoms)] * dim).T
+    assert (offsets.shape == (n_atoms, dim))
+    for i in range(dim):
+        onp.random.shuffle(offsets[:, i])
+    conf = (0.5 * randn(n_atoms, dim) + offsets)
+    conf += conf.min()
+    params = rand(n_atoms, 3)
+    params[:, 0] -= np.mean(params[:, 0])
+
+    box = np.diag(0.5 + 0.5 * rand(3)) * 10 * n_atoms
+
+    lamb = rand()
+
+    charge_rescale_mask = onp.ones((n_atoms, n_atoms))
+    lj_rescale_mask = onp.ones((n_atoms, n_atoms))
+    for _ in range(n_atoms):
+        i, j = randint(n_atoms, size=2)
+        charge_rescale_mask[i, j] = 0.0
+        lj_rescale_mask[i, j] = 0.0
+
+    beta = rand() + 1
+    cutoff = rand() + 0.5
+
+    lambda_plane_idxs = randint(low=-2, high=2, size=n_atoms)
+    lambda_offset_idxs = randint(low=-2, high=2, size=n_atoms)
+
+    args = conf, params, box, lamb, charge_rescale_mask, lj_rescale_mask, beta, cutoff, lambda_plane_idxs, lambda_offset_idxs
+
+    return args
+
+
+def compare_two_potentials(u_a: NonbondedFxn, u_b: NonbondedFxn, args: NonbondedArgs, differentiate_wrt=(0, 1, 3)):
+    """Assert that energies and derivatives w.r.t. request argnums are close"""
+    value_and_grads = partial(value_and_grad, argnums=differentiate_wrt)
+    energy_a, gradients_a = value_and_grads(u_a)(*args)
+    energy_b, gradients_b = value_and_grads(u_b)(*args)
+
+    assert energy_a == energy_b
+    onp.testing.assert_almost_equal(energy_a, energy_b)
```

Comment:
you probably want to specify the tolerances for these based on the precision of the method, I think the defaults are for float32.

Reply:
I'm not sure how to set the precision of a Jax function. Could you point me to how to do this? (I thought Jax default precision is set globally)

Reply:
Hm, so far we've never used the 32-bit version of jax, since we use it mainly as a *reference* platform where speed isn't of the paramount importance. That is, previously, we've always compared 32-bit and 64-bit custom_ops against *only* the 64bit version of jax. I think comparing the 64-bit nonbonded_v3 vs the 64-bit nonbonded_v3_specific pairs is probably sufficient, but you're right that there's no easy way to test the 32-bit versions except to globally enable it in another file.